src/library/autodj/autodjfeature.cpp

```diff
@@ -275,41 +271,61 @@ void AutoDJFeature::slotCrateAutoDjChanged(int crateId, bool added) {
             }
         }
     }
-    // Let subscribers know whether it's possible to add a random track.
-    emit(enableAddRandom(m_crateList.length() > 0));
 #endif // __AUTODJCRATES__
 }
+// Adds a random track : this will be faster when there are sufficiently large 
+// tracks in the crates
 
 void AutoDJFeature::slotAddRandomTrack(bool) {
 #ifdef __AUTODJCRATES__
-    int retriveAttempts = 0;
-    while (retriveAttempts < kMaxRetiveAttempts) {
-        // Get access to the auto-DJ playlist.
-        PlaylistDAO& playlistDao = m_pTrackCollection->getPlaylistDAO();
-        if (m_iAutoDJPlaylistId >= 0) {
+    int failedRetriveAttempts = 0;
+    bool playlistHasTrack = false;
+
+    // Get access to the auto-DJ playlist
+    PlaylistDAO& playlistDao = m_pTrackCollection->getPlaylistDAO();
+    int iTrackId = -1;
+
+    if (m_iAutoDJPlaylistId >= 0) {
+        while (failedRetriveAttempts < kMaxRetriveAttempts) {
             // Get the ID of a randomly-selected track.
-            int iTrackId = m_autoDjCratesDao.getRandomTrackId();
-            //qDebug() << "we got iTrackid " <<iTrackId;
+            iTrackId = m_autoDjCratesDao.getRandomTrackId();
             if (iTrackId != -1) {
                 // Get Track Information
                 TrackPointer addedTrack = (m_pTrackCollection->getTrackDAO()).getTrack(iTrackId);
-                if(addedTrack->exists()) {
-                    //qDebug() << addedTrack->getTitle() << " Exists";
-                    // Add this randomly-selected track to the auto-DJ playlist.
+                playlistHasTrack = playlistDao.playlistHasTrack(iTrackId, m_iAutoDJPlaylistId);
+                if(addedTrack->exists() && !playlistHasTrack) {
```

Comment:
Im not sure I understand this. iTrackId gets a random track which then is passed to 
````
 TrackPointer addedTrack = (m_pTrackCollection->getTrackDAO()).getTrack(iTrackId);
````
then 
````
addedTrack->exists() ;
```` 
checks if the track exists. Further the track Id and autoDJ playlist ID is passed to  playListHasTrack here,
````
playlistHasTrack = playlistDao.playlistHasTrack(iTrackId, m_iAutoDJPlaylistId);
````
Which returns true if the autoDJ playlist has the track . In which case the loop continues to attempt to retrieve another track. Is this not the check you asked for?